src/server.rs

```diff
@@ -469,4 +473,112 @@ pub extern "C" fn rustls_server_session_get_sni_hostname(
         }
         rustls_result::Ok
     }
-}
\ No newline at end of file
+}
+
+/// The TLS Client Hello information provided to a ClientHelloCallback function.
+/// `sni_name` is the SNI servername provided by the client (not 0 terminated) with
+/// `sni_name_len` as its length. If the client did not provide an SNI, the name
+/// length is 0.
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_client_hello {
+    sni_name: *const c_char,
+    sni_name_len: size_t,
+    signature_schemes: *const c_ushort,
+    signature_schemes_len: size_t,
+}
+
+/// Any context information the callback will receive when invoked.
+type ClientHelloUserData = *mut c_void;
+
+/// Called when the TLS ClientHello message was received and successfully decoded.
+type ClientHelloCallback = unsafe extern "C" fn(
+    userdata: ClientHelloUserData,
+    hello: *const rustls_client_hello,
+) -> rustls_result;
+
+struct ClientHelloResolver {
+    /// Implementation of rustls::ResolvesServerCert that passes values
+    /// from the supplied ClientHello to the callback function.
+    pub callback: ClientHelloCallback,
+    pub userdata: ClientHelloUserData,
+    pub previous_resolver: Option<Arc<dyn rustls::ResolvesServerCert>>,
+}
+
+impl ClientHelloResolver {
+    pub fn new(
+        callback: ClientHelloCallback,
+        userdata: ClientHelloUserData,
+        previous_resolver: Option<Arc<dyn rustls::ResolvesServerCert>>,
+    ) -> ClientHelloResolver {
+        ClientHelloResolver { callback, userdata, previous_resolver }
+    }
+}
+
+impl rustls::ResolvesServerCert for ClientHelloResolver {
+    fn resolve(&self, client_hello: ClientHello<'_>) -> Option<CertifiedKey> {
+        let sni_name = {
```

Comment:
```suggestion
        let sni_name: &str = {
```

I prefer lots of type annotations, particularly when relying on type inference like `c.into()`.